fedimint-core/src/bls12_381_serde.rs

```diff
@@ -1,81 +1,74 @@
 pub mod scalar {
-    use bitcoin_hashes::hex::{FromHex, ToHex};
     use bls12_381::Scalar;
     use serde::de::Error;
-    use serde::{Deserialize, Deserializer, Serializer};
+    use serde::{Deserializer, Serializer};
 
-    pub fn serialize<S: Serializer>(x: &Scalar, s: S) -> Result<S::Ok, S::Error> {
-        let bytes = x.to_bytes();
-
-        if s.is_human_readable() {
-            s.serialize_str(&bytes.to_hex())
-        } else {
-            s.serialize_bytes(&bytes)
-        }
+    pub fn serialize<S>(scalar: &Scalar, s: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serdect::array::serialize_hex_lower_or_bin(&scalar.to_bytes(), s)
     }
 
-    pub fn deserialize<'d, D: Deserializer<'d>>(d: D) -> Result<Scalar, D::Error> {
-        let bytes: Vec<u8> = if d.is_human_readable() {
-            let deser: String = Deserialize::deserialize(d)?;
-            Vec::<u8>::from_hex(&deser).map_err(serde::de::Error::custom)?
-        } else {
-            Deserialize::deserialize(d)?
-        };
-
-        if bytes.len() != 32 {
-            return Err(Error::invalid_length(bytes.len(), &"32 bytes"));
-        }
-
-        let mut byte_array = [0u8; 32];
+    pub fn deserialize<'d, D>(d: D) -> Result<Scalar, D::Error>
+    where
+        D: Deserializer<'d>,
+    {
+        let mut byte_array = [0; 32];
 
-        byte_array.copy_from_slice(&bytes);
+        serdect::array::deserialize_hex_or_bin(&mut byte_array, d)?;
 
         Option::from(Scalar::from_bytes(&byte_array))
-            .ok_or_else(|| Error::custom("Could not decode compressed G1Affine"))
+            .ok_or(Error::custom("Could not decode scalar"))
```

Comment:
`Error::custom` is a heavy-ish operation (it allocates), should be only run on-demand by keeping `ok_or_else`.

![Screenshot 2024-03-29 at 21-18-49 value rs - source](https://github.com/fedimint/fedimint/assets/84478054/d28abdf6-919e-464c-875d-e3b710ce64a0)